File: memorezar-android/app/src/main/java/com/memorezar/app/data/services/PurchaseService.kt
package com.memorezar.app.data.services

import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.memorezar.app.core.alert.SoundTheme
import com.memorezar.app.data.models.SuggestionPack
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Date
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "PurchaseService"
private const val INSTALL_DATE_KEY = "memorezar_first_install"

/**
 * Manages Pro subscriptions and feature gating via RevenueCat.
 *
 * NOTE: RevenueCat Android SDK integration is stubbed here.
 * To activate, add the RevenueCat dependency and uncomment the SDK calls.
 * The entitlement/feature gating logic is fully implemented.
 */
@Singleton
class PurchaseService @Inject constructor(
    @ApplicationContext private val appContext: Context
) {
    private val prefs: SharedPreferences =
        appContext.getSharedPreferences("memorezar_purchases", Context.MODE_PRIVATE)

    private val _isPro = MutableStateFlow(false)
    val isPro: StateFlow<Boolean> = _isPro.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    val installDate: Date by lazy {
        val stored = prefs.getLong(INSTALL_DATE_KEY, 0L)
        if (stored > 0) {
            Date(stored)
        } else {
            val now = Date()
            prefs.edit().putLong(INSTALL_DATE_KEY, now.time).apply()
            now
        }
    }

    // Early adopter: currently everyone gets full access
    val isEarlyAdopter: Boolean get() = true  // earlyAdopterCutoff = distantFuture

    val hasFullAccess: Boolean get() = _isPro.value || isEarlyAdopter

    // MARK: - Configuration

    fun configure() {
        // TODO: Initialize RevenueCat SDK
        // Purchases.configure(PurchasesConfiguration.Builder(appContext, RevenueCatConfig.API_KEY).build())
        // refreshCustomerInfo()
        // fetchOfferings()

        // Record install date on first launch
        installDate // triggers lazy init
        Log.d(TAG, "PurchaseService configured, installDate=$installDate")
    }

    // MARK: - User Identity

    fun identifyUser(userId: String) {
        // TODO: Purchases.sharedInstance.logIn(userId) { ... }
        Log.d(TAG, "Identifying user: $userId")
    }

    fun logOutUser() {
        // TODO: Purchases.sharedInstance.logOut()
        _isPro.value = false
        Log.d(TAG, "User logged out from RevenueCat")
    }

    // MARK: - Purchases

    suspend fun purchasePro(activity: Activity) {
        // TODO: Implement RevenueCat purchase flow
        // val offerings = Purchases.sharedInstance.getOfferings()
        // val package = offerings.current?.availablePackages?.firstOrNull()
        // Purchases.sharedInstance.purchase(PurchaseParams.Builder(activity, package).build())
        Log.d(TAG, "Purchase flow not yet implemented")
    }

    suspend fun restorePurchases() {
        // TODO: Purchases.sharedInstance.restorePurchases()
        Log.d(TAG, "Restore purchases not yet implemented")
    }

    // MARK: - Feature Gating

    fun canAddCustomQuote(currentCount: Int): Boolean {
        if (hasFullAccess) return true
        return currentCount < 20
    }

    fun canAddRecording(currentCount: Int): Boolean {
        if (hasFullAccess) return true
        return currentCount < 3
    }

    val canUseTTS: Boolean get() = hasFullAccess

    fun canUseSoundTheme(theme: SoundTheme): Boolean {
        if (hasFullAccess) return true
        return theme == SoundTheme.DEFAULT
    }

    fun canAccessPack(pack: SuggestionPack): Boolean {
        if (hasFullAccess) return true
        return pack.isFree
    }

    // MARK: - Internal

    private fun refreshCustomerInfo() {
        // TODO: Purchases.sharedInstance.getCustomerInfo { info, error ->
        //     _isPro.value = info?.entitlements?.get(RevenueCatConfig.PRO_ENTITLEMENT_ID)?.isActive == true
        // }
    }
}

File: ios/Memorezar/Data/Services/PurchaseService.swift
import Foundation
import RevenueCat
import SwiftUI

/// Manages in-app purchases and entitlements via RevenueCat.
/// Singleton following the same pattern as AuthService.
///
/// **Current state:** All features are free. Gating will be added later.
/// Install date is recorded silently so early adopters can be grandfathered.
final class PurchaseService: NSObject, ObservableObject {
    static let shared = PurchaseService()

    // MARK: - Published State

    @Published var isPro: Bool = false
    @Published var offerings: Offerings?

    // MARK: - Early Adopter

    /// Users who installed before this date get Pro free forever.
    /// Set to .distantFuture while the app is free for everyone.
    /// When monetizing, change to the cutoff date.
    static let earlyAdopterCutoff: Date = .distantFuture

    /// First install date, recorded once and persisted forever.
    var installDate: Date {
        let key = "memorezar_first_install"
        if let stored = UserDefaults.standard.object(forKey: key) as? Date {
            return stored
        }
        let now = Date()
        UserDefaults.standard.set(now, forKey: key)
        return now
    }

    var isEarlyAdopter: Bool {
        installDate < Self.earlyAdopterCutoff
    }

    /// Whether the user has full access (Pro subscriber OR early adopter).
    var hasFullAccess: Bool {
        isPro || isEarlyAdopter
    }

    // MARK: - Init

    private override init() {
        super.init()
    }

    // MARK: - Configuration

    /// Call once at app launch before any purchases.
    func configure() {
        // Record install date on first launch
        _ = installDate

        Purchases.configure(withAPIKey: RevenueCatConfig.apiKey)
        Purchases.shared.delegate = self

        Task {
            await refreshCustomerInfo()
            await fetchOfferings()
        }
    }

    // MARK: - User Identification

    /// Link RevenueCat to the Supabase user on sign-in.
    func identifyUser(userId: String) {
        Purchases.shared.logIn(userId) { [weak self] customerInfo, _, _ in
            if let info = customerInfo {
                self?.updateEntitlements(from: info)
            }
        }
    }

    /// Revert to anonymous RevenueCat user on sign-out.
    func logOutUser() {
        Purchases.shared.logOut { [weak self] customerInfo, _ in
            if let info = customerInfo {
                self?.updateEntitlements(from: info)
            }
        }
    }

    // MARK: - Purchases

    /// Purchase a Pro subscription or lifetime.
    @MainActor
    func purchasePro(package: Package) async throws {
        let result = try await Purchases.shared.purchase(package: package)
        updateEntitlements(from: result.customerInfo)
    }

    /// Restore previous purchases.
    @MainActor
    func restorePurchases() async throws {
        let info = try await Purchases.shared.restorePurchases()
        updateEntitlements(from: info)
    }

    // MARK: - Entitlement Checks (for future gating)
    //
    // These all return true right now because hasFullAccess is true
    // for everyone (earlyAdopterCutoff = .distantFuture).
    // When monetizing, change the cutoff date and these will gate properly.

    func canAddCustomQuote(quoteStore: QuoteStore) -> Bool {
        if hasFullAccess { return true }
        let customCount = quoteStore.quotes.filter { quote in
            let category = quoteStore.categories.first { $0.id == quote.categoryId }
            return category?.sourcePackId == nil
        }.count
        return customCount < 20
    }

    func canAddRecording(recordingStore: LocalRecordingStore) -> Bool {
        if hasFullAccess { return true }
        let userRecordingCount = recordingStore.recordings.filter { $0.sourceRecordingId == nil }.count
        return userRecordingCount < 3
    }

    var canUseTTS: Bool { hasFullAccess }

    func canUseSoundTheme(_ theme: SoundTheme) -> Bool {
        hasFullAccess || theme == .default
    }

    /// Pro users get all packs. Free users only see packs marked as free.
    func canAccessPack(_ pack: SuggestionPack) -> Bool {
        hasFullAccess || pack.isFree
    }

    // MARK: - Private Helpers

    private func fetchOfferings() async {
        do {
            let offers = try await Purchases.shared.offerings()
            await MainActor.run { self.offerings = offers }
        } catch {
            print("[PurchaseService] Failed to fetch offerings: \(error)")
        }
    }

    private func refreshCustomerInfo() async {
        do {
            let info = try await Purchases.shared.customerInfo()
            await MainActor.run { self.updateEntitlements(from: info) }
        } catch {
            print("[PurchaseService] Failed to refresh customer info: \(error)")
        }
    }

    private func updateEntitlements(from info: CustomerInfo) {
        DispatchQueue.main.async {
            self.isPro = info.entitlements[RevenueCatConfig.proEntitlementId]?.isActive == true
        }
    }
}

// MARK: - PurchasesDelegate

extension PurchaseService: PurchasesDelegate {
    func purchases(_ purchases: Purchases, receivedUpdated customerInfo: CustomerInfo) {
        updateEntitlements(from: customerInfo)
    }
}
